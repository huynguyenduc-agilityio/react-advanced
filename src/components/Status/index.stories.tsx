import type { Meta, StoryObj } from '@storybook/react';

// Types
import { Status as StatusEnum } from '@/types';

// Components
import Status from '.';

const meta = {
  title: 'Components/Status',
  component: Status,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Status>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Online: Story = {
  args: {
    type: StatusEnum.Online,
  },
};

export const Offline: Story = {
  args: {
    type: StatusEnum.Offline,
  },
};
