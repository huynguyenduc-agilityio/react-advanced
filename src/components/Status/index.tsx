import { Flex, Text } from '@chakra-ui/react';

// Icons
import { EllipseIcon } from '@/icons';

// Enums
import { Status as StatusEnum } from '@/enums';

export interface StatusProps {
  type?: StatusEnum;
}

const Status = ({ type = StatusEnum.Online }: StatusProps) => {
  return (
    <Flex
      maxW="60px"
      alignItems="center"
      justifyContent="center"
      gap={1}
      px={2}
      py={1}
      borderRadius="2px"
      border="1px solid"
      bg={type === StatusEnum.Online ? '#09363B' : '#39425F'}
      borderColor={type === StatusEnum.Online ? '#067B51' : '#515A79'}
    >
      <EllipseIcon
        color={(type !== StatusEnum.Online && '#AEB9E1') || undefined}
      />
      <Text
        fontWeight="medium"
        fontSize="xs"
        color={type === StatusEnum.Online ? '#14CA74' : 'pastelBlue'}
      >
        {type === StatusEnum.Online ? 'Online' : 'Offline'}
      </Text>
    </Flex>
  );
};

export default Status;
