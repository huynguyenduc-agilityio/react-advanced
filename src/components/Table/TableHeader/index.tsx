import { memo } from 'react';
import isEqual from 'react-fast-compare';
import { Checkbox, Flex, Icon, Text, Th, Tr } from '@chakra-ui/react';

// Types
import { THeaderTable } from '@/types';

interface TableHeadProps {
  isAllSelected: boolean;
  isIndeterminate: boolean;
  columns?: THeaderTable[];
  onSelectAll: (selected: boolean) => void;
}

const TableHeader = ({
  columns = [],
  isAllSelected,
  isIndeterminate,
  onSelectAll,
}: TableHeadProps) => (
  <Tr>
    <Th py={4} pl={9} pr={0} textAlign="center">
      <Flex>
        <Checkbox
          aria-label="Checkbox Indeterminate"
          isChecked={isAllSelected}
          isIndeterminate={isIndeterminate}
          onChange={(e) => onSelectAll(e.target.checked)}
        />
      </Flex>
    </Th>

    {!!columns.length &&
      columns.map(({ key, title, icon }) => (
        <Th key={key} py={4} textAlign="left" fontFamily="MonaSans-Regular">
          <Flex alignItems="center" gap={1}>
            {icon && <Icon as={icon} boxSize={3} color="pastelBlue" />}
            <Text
              color="white"
              textTransform="none"
              fontWeight="semibold"
              title={title}
            >
              {title}
            </Text>
          </Flex>
        </Th>
      ))}
  </Tr>
);

export default memo(TableHeader, isEqual);
