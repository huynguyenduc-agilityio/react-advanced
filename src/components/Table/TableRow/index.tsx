import { memo } from 'react';
import isEqual from 'react-fast-compare';
import { Checkbox, Flex, Td, Tr } from '@chakra-ui/react';

// Types
import { TCellTable } from '@/types';

interface TableRowProps {
  id: string;
  cells: TCellTable[];
  isSelected: boolean;
  onSelect: (rowId: string) => void;
}

const TableRow = ({ id, cells = [], isSelected, onSelect }: TableRowProps) => {
  const handleSelectRecord = () => {
    onSelect(id);
  };

  return (
    <Tr>
      <Td py={4} pl={9} pr={0} textAlign="center">
        <Flex>
          <Checkbox
            aria-label={`Checkbox ${id}`}
            isChecked={isSelected}
            onChange={handleSelectRecord}
          />
        </Flex>
      </Td>
      {cells.map((cell) => (
        <Td
          key={cell.key}
          py={4}
          textAlign="left"
          fontSize="xs"
          color="pastelBlue"
        >
          {cell.content}
        </Td>
      ))}
    </Tr>
  );
};

export default memo(TableRow, isEqual);
