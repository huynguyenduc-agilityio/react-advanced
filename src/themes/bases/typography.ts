import { ThemeOverride } from '@chakra-ui/react';

export const fonts = {
  body: "'MonaSans-Regular', sans-serif",
  heading: "'WorkSans-SemiBold', sans-serif",
};

export const fontSizes: ThemeOverride['fontSizes'] = {
  xs: '10px',
  sm: '12px',
  md: '14px',
  lg: '16px',
  xl: '20px',
  '2xl': '24px',
};
