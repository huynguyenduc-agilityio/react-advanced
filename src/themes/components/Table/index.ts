import { ComponentStyleConfig } from '@chakra-ui/react';

export const Table: ComponentStyleConfig = {
  baseStyle: {
    table: {
      bg: 'midNightBlue',
      th: { border: 'none' },
      td: { border: 'none' },
    },
  },
  variants: {
    striped: {
      tbody: {
        tr: {
          '&:nth-of-type(odd) td': {
            background: 'darkBlue',
          },
        },
      },
    },
  },
  defaultProps: {
    variant: 'striped',
  },
};
