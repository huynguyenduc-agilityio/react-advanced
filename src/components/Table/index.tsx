import {
  Button,
  Flex,
  Table as TableChakra,
  TableContainer,
  TableProps,
  Tbody,
  Td,
  Text,
  Thead,
  Tr,
} from '@chakra-ui/react';
import { memo, useCallback, useMemo, useState } from 'react';

// Types
import { TDataSource, THeaderTable, ToastStatus } from '@/types';

// Constants
import { ERROR_MESSAGES, SUCCESS_MESSAGES } from '@/constants';

// Utils
import { processTableData } from '@/utils';

// Hooks
import { useCustomToast, useDeleteMultipleUsers } from '@/hooks';

// Contexts
import { useConfirmationDialog } from '@/contexts';

// Components
import Fallback from '@/components/Fallback';
import TableHeader from '@/components/Table/TableHeader';
import TableRow from '@/components/Table/TableRow';

type TTableProps = TableProps & {
  title?: string;
  isLoading?: boolean;
  columns?: THeaderTable[];
  dataSource?: TDataSource[];
};

const Table = ({
  title = 'All Users',
  isLoading,
  columns = [],
  dataSource = [],
  ...props
}: TTableProps) => {
  const processedData = useMemo(
    () => processTableData(columns, dataSource),
    [columns, dataSource],
  );
  const showToast = useCustomToast();
  const { confirm } = useConfirmationDialog();
  const { handleDeleteMultipleUsers, isDeleteMultipleLoading } =
    useDeleteMultipleUsers();

  // State to track selected checkboxes
  const [selectedRows, setSelectedRows] = useState<Set<string>>(new Set());
  const [isAllSelected, setIsAllSelected] = useState(false);

  // Handle single row selection
  const handleRowSelect = useCallback(
    (rowId: string) => {
      const updatedSelectedRows = new Set(selectedRows);

      if (updatedSelectedRows.has(rowId)) {
        updatedSelectedRows.delete(rowId);
      } else {
        updatedSelectedRows.add(rowId);
      }

      setSelectedRows(updatedSelectedRows);
      setIsAllSelected(updatedSelectedRows.size === processedData.length);
    },
    [processedData.length, selectedRows],
  );

  // Handle "Select All" checkbox
  const handleSelectAll = useCallback(
    (selected: boolean) => {
      if (selected) {
        const allRowIds = processedData.map((data) => data.id);
        setSelectedRows(new Set(allRowIds));
      } else {
        setSelectedRows(new Set());
      }
      setIsAllSelected(selected);
    },
    [processedData],
  );

  const handleBulkDelete = useCallback(async () => {
    try {
      await handleDeleteMultipleUsers(Array.from(selectedRows));

      setSelectedRows(new Set());
      setIsAllSelected(false);

      showToast({
        status: ToastStatus.Success,
        message: SUCCESS_MESSAGES.DELETE_USER,
      });
    } catch (error) {
      showToast({
        status: ToastStatus.Error,
        message: ERROR_MESSAGES.DELETE_USER,
      });
    }
  }, [selectedRows, handleDeleteMultipleUsers, showToast]);

  const handleConfirmBulkDelete = useCallback(() => {
    confirm({
      title: 'Delete Users',
      confirmMessage: 'Are you sure you want to delete users?',
      onConfirm: handleBulkDelete,
    });
  }, [confirm, handleBulkDelete]);

  // Determine if the "Select All" checkbox is in indeterminate state
  const isIndeterminate =
    selectedRows.size > 0 && selectedRows.size < processedData.length;

  return (
    <TableContainer
      w="full"
      h="full"
      borderRadius="12px"
      border="0.6px solid"
      borderColor="slateBlue"
      overflowY="scroll"
      css={{
        '&::-webkit-scrollbar': {
          width: 2,
        },
        '&::-webkit-scrollbar-track': {
          width: 2,
        },
        '&::-webkit-scrollbar-thumb': {
          background: 'gray',
          borderRadius: '24px',
        },
      }}
    >
      {isDeleteMultipleLoading && (
        <Flex
          position="absolute"
          inset={0}
          bg="rgba(0, 0, 0, 0.6)"
          alignItems="center"
          justifyContent="center"
          zIndex={10}
        >
          <Fallback />
        </Flex>
      )}

      <Flex
        alignItems="center"
        justifyContent="space-between"
        w="full"
        px={9}
        py={6}
        bg="midNightBlue"
        borderBottom="0.6px solid"
        borderColor="slateBlue"
      >
        <Text as="header" fontFamily="WorkSans-Medium" color="white">
          {title}
        </Text>
        {selectedRows.size !== 0 && (
          <Button onClick={handleConfirmBulkDelete}>Mark Delete</Button>
        )}
      </Flex>

      <TableChakra variant="striped" {...props}>
        <Thead>
          <TableHeader
            columns={columns}
            onSelectAll={handleSelectAll}
            isAllSelected={isAllSelected}
            isIndeterminate={isIndeterminate}
          />
        </Thead>

        <Tbody>
          {processedData.map((data) => (
            <TableRow
              key={data.id}
              id={data.id}
              cells={data.cells}
              isSelected={selectedRows.has(data.id)}
              onSelect={handleRowSelect}
            />
          ))}

          {!processedData.length && !isLoading && (
            <Tr>
              <Td
                colSpan={columns?.length + 1} // Add +1 for the checkbox column
                textAlign="center"
                fontWeight="semibold"
                border="none"
                fontSize="md"
              >
                {ERROR_MESSAGES.EMPTY_DATA}
              </Td>
            </Tr>
          )}

          {!!processedData.length && isLoading && (
            <Tr position="absolute" top="50%" left="50%">
              <Td
                colSpan={columns?.length + 1}
                width="full"
                textAlign="center"
                fontWeight="semibold"
                border="none"
                fontSize="md"
              >
                <Fallback />
              </Td>
            </Tr>
          )}

          {!processedData.length && isLoading && (
            <Tr>
              <Td
                colSpan={columns?.length + 1}
                width="full"
                textAlign="center"
                fontWeight="semibold"
                border="none"
                fontSize="md"
              >
                <Fallback />
              </Td>
            </Tr>
          )}
        </Tbody>
      </TableChakra>
    </TableContainer>
  );
};

export default memo(Table);
