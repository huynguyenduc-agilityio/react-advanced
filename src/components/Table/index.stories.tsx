import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import type { Meta, StoryObj } from '@storybook/react';
import { Avatar, Box, Flex, IconButton, Text } from '@chakra-ui/react';

// Icons
import { FaPhoneAlt, FaUser } from 'react-icons/fa';
import { HiMapPin, HiMiniPencil } from 'react-icons/hi2';
import { IoBagSharp } from 'react-icons/io5';
import { PiCheckSquareFill } from 'react-icons/pi';
import { RiDeleteBin7Fill } from 'react-icons/ri';

// Components
import { Status } from '@/components';
import Table from '.';

// Enums
import { Status as StatusEnum } from '@/enums';

// Types
import { TDataSource } from '@/types';

// Themes
import { fonts } from '@/themes/bases/typography';

const queryClient = new QueryClient();

const meta = {
  title: 'Components/Table',
  component: Table,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  render: (args) => (
    <QueryClientProvider client={queryClient}>
      <Box w="100%" h="screen">
        <Table {...args} />
      </Box>
    </QueryClientProvider>
  ),
} satisfies Meta<typeof Table>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    columns: [
      {
        title: 'Name',
        key: 'name',
        icon: FaUser,
        renderBody: ({ name, avatar, email }: TDataSource) => (
          <Flex p={0} gap={1} border="none" fontSize="xs" alignItems="center">
            <Avatar size="sm" name={name as string} src={avatar as string} />
            <Box>
              <Text
                fontSize="xs"
                fontWeight="medium"
                fontFamily={fonts.heading}
                color="white"
              >
                {name}
              </Text>
              <Text fontSize="xs" fontWeight="medium" color="pastelBlue">
                {email}
              </Text>
            </Box>
          </Flex>
        ),
      },
      {
        title: 'Phone',
        key: 'phone',
        icon: FaPhoneAlt,
      },
      {
        title: 'Location',
        key: 'location',
        icon: HiMapPin,
      },
      {
        title: 'Company',
        key: 'company',
        icon: IoBagSharp,
      },
      {
        title: 'Status',
        key: 'status',
        icon: PiCheckSquareFill,
        renderBody: ({ status }) => <Status type={status as StatusEnum} />,
      },
      {
        key: 'action',
        renderBody: () => (
          <Flex gap={2}>
            <IconButton
              aria-label="Edit"
              icon={<HiMiniPencil />}
              variant="icon"
            />
            <IconButton
              aria-label="delete"
              icon={<RiDeleteBin7Fill />}
              variant="icon"
            />
          </Flex>
        ),
      },
    ],
    dataSource: [
      {
        id: '1',
        avatar: 'https://bit.ly/dan-abramov',
        name: 'John Caster',
        email: 'john@google.com',
        phone: '(414) 907 - 1274',
        location: 'United States',
        company: 'Google',
        status: 'online',
      },
      {
        id: '2',
        avatar: 'https://bit.ly/dan-abramov',
        name: 'John Caster 2',
        email: 'john@google.com',
        phone: '(414) 907 - 1274',
        location: 'United States',
        company: 'Google',
        status: 'offline',
      },
      {
        id: '3',
        avatar: 'https://bit.ly/dan-abramov',
        name: 'John Caster 3',
        email: 'john@google.com',
        phone: '(414) 907 - 1274',
        location: 'United States',
        company: 'Google',
        status: 'online',
      },
    ],
  },
};
