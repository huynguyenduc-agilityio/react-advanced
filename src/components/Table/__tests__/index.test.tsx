import { ReactNode } from 'react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { Avatar, Box, Flex, IconButton, Text } from '@chakra-ui/react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Icons
import { FaPhoneAlt, FaUser } from 'react-icons/fa';
import { HiMapPin, HiMiniPencil } from 'react-icons/hi2';
import { IoBagSharp } from 'react-icons/io5';
import { PiCheckSquareFill } from 'react-icons/pi';
import { RiDeleteBin7Fill } from 'react-icons/ri';

// Themes
import { fonts } from '@/themes/bases/typography';

// Types
import { TDataSource, Status as StatusEnum } from '@/types';

// Components
import Status from '@/components/Status';
import Table from '..';

const mockOnClick = jest.fn();
const mockDelete = jest.fn();

jest.mock('@/hooks', () => ({
  ...jest.requireActual('@/hooks'),
  useDeleteMultipleUsers: jest.fn(() => ({
    handleDeleteMultipleUsers: mockDelete,
    isDeleteMultipleLoading: false,
  })),
}));

jest.mock('@/contexts', () => ({
  ...jest.requireActual('@/contexts'),
  useConfirmationDialog: jest.fn(() => ({
    confirm: jest.fn(({ onConfirm }) => onConfirm()),
  })),
}));

describe('Table Component', () => {
  const queryClient = new QueryClient();

  const Wrapper = ({ children }: { children: ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  const dataSource: TDataSource[] = [
    {
      id: '1',
      name: 'John Doe',
      phone: '123-456-7890',
      location: 'New York, NY',
      company: 'ABC Company',
      status: StatusEnum.Online,
    },
    {
      id: '2',
      name: 'Jane Doe',
      phone: '987-654-3210',
      location: 'Los Angeles, CA',
      company: 'XYZ Company',
      status: StatusEnum.Offline,
    },
  ];

  const renderSetup = () =>
    render(
      <Wrapper>
        <Table
          columns={[
            {
              title: 'Name',
              key: 'name',
              icon: FaUser,
              renderBody: ({ name, avatar, email }: TDataSource) => (
                <Flex
                  p={0}
                  gap={1}
                  border="none"
                  fontSize="xs"
                  alignItems="center"
                >
                  <Avatar
                    size="sm"
                    name={name as string}
                    src={avatar as string}
                  />
                  <Box>
                    <Text
                      fontSize="xs"
                      fontWeight="medium"
                      fontFamily={fonts.heading}
                      color="white"
                    >
                      {name}
                    </Text>
                    <Text fontSize="xs" fontWeight="medium" color="pastelBlue">
                      {email}
                    </Text>
                  </Box>
                </Flex>
              ),
            },
            {
              title: 'Phone',
              key: 'phone',
              icon: FaPhoneAlt,
            },
            {
              title: 'Location',
              key: 'location',
              icon: HiMapPin,
            },
            {
              title: 'Company',
              key: 'company',
              icon: IoBagSharp,
            },
            {
              title: 'Status',
              key: 'status',
              icon: PiCheckSquareFill,
              renderBody: ({ status }) => (
                <Status type={status as StatusEnum} />
              ),
            },
            {
              key: 'action',
              renderBody: () => (
                <Flex gap={2}>
                  <IconButton
                    aria-label="Edit"
                    icon={<HiMiniPencil />}
                    variant="icon"
                    data-testid="edit-button"
                    onClick={mockOnClick}
                  />
                  <IconButton
                    aria-label="Delete"
                    icon={<RiDeleteBin7Fill />}
                    variant="icon"
                  />
                </Flex>
              ),
            },
          ]}
          dataSource={dataSource}
        />
      </Wrapper>,
    );

  it('should match snapshot', () => {
    const { container } = renderSetup();

    expect(container).toMatchSnapshot();
  });

  it('should render an empty state when no data is provided ', () => {
    const { getByText } = render(
      <Wrapper>
        <Table />
      </Wrapper>,
    );

    expect(getByText('No data found')).toBeInTheDocument();
  });

  it('should render no data correctly', () => {
    const { getByText } = render(
      <Wrapper>
        <Table columns={[]} dataSource={[]} />
      </Wrapper>,
    );

    expect(getByText('No data found')).toBeInTheDocument();
  });

  it('onClick edit button should be called', async () => {
    const { getAllByTestId } = renderSetup();
    const editButtons = getAllByTestId('edit-button');

    await userEvent.click(editButtons[0]);
    expect(mockOnClick).toHaveBeenCalled();
  });

  it('should allow selecting a row', async () => {
    const { getAllByRole } = renderSetup();
    const checkboxes = getAllByRole('checkbox');

    await userEvent.click(checkboxes[1]);

    expect(checkboxes[1]).toBeChecked();
  });

  it('should allow selecting all rows', async () => {
    const { getAllByRole } = renderSetup();
    const checkboxes = getAllByRole('checkbox');

    await userEvent.click(checkboxes[0]);

    checkboxes.forEach((checkbox) => {
      expect(checkbox).toBeChecked();
    });
  });

  it('should call delete action on confirm', async () => {
    const { getByText, getAllByRole } = renderSetup();

    const checkboxes = getAllByRole('checkbox');

    await userEvent.click(checkboxes[1]);
    await userEvent.click(getByText('Mark Delete'));

    expect(mockDelete).toHaveBeenCalled();
  });

  it('should call delete action on confirm', async () => {
    const { getAllByTestId, getByText } = renderSetup();

    const checkboxes = getAllByTestId('row-checkbox');
    await userEvent.click(checkboxes[0]);

    const deleteButton = getByText('Mark Delete');
    expect(deleteButton).toBeInTheDocument();

    await userEvent.click(deleteButton);

    expect(mockDelete).toHaveBeenCalled();
  });
});
