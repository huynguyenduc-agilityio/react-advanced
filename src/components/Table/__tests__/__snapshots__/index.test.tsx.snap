// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should match snapshot 1`] = `
<div>
  <div
    class="chakra-table__container css-zvwuj9"
  >
    <div
      class="css-h6v4s3"
    >
      <header
        class="chakra-text css-oerhs"
      >
        All Users
      </header>
    </div>
    <table
      class="chakra-table css-0"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-zgnwis"
          >
            <div
              class="css-k008qs"
            >
              <label
                class="chakra-checkbox css-1elxtiq"
              >
                <input
                  aria-label="Checkbox Indeterminate"
                  class="chakra-checkbox__input"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="checkbox"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="chakra-checkbox__control css-dnty2r"
                />
              </label>
            </div>
          </th>
          <th
            class="css-ykdcwi"
          >
            <div
              class="css-1qofolz"
            >
              <svg
                class="chakra-icon css-1baa01z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                />
              </svg>
              <p
                class="chakra-text css-qrr9p8"
                title="Name"
              >
                Name
              </p>
            </div>
          </th>
          <th
            class="css-ykdcwi"
          >
            <div
              class="css-1qofolz"
            >
              <svg
                class="chakra-icon css-1baa01z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M497.39 361.8l-112-48a24 24 0 0 0-28 6.9l-49.6 60.6A370.66 370.66 0 0 1 130.6 204.11l60.6-49.6a23.94 23.94 0 0 0 6.9-28l-48-112A24.16 24.16 0 0 0 122.6.61l-104 24A24 24 0 0 0 0 48c0 256.5 207.9 464 464 464a24 24 0 0 0 23.4-18.6l24-104a24.29 24.29 0 0 0-14.01-27.6z"
                />
              </svg>
              <p
                class="chakra-text css-qrr9p8"
                title="Phone"
              >
                Phone
              </p>
            </div>
          </th>
          <th
            class="css-ykdcwi"
          >
            <div
              class="css-1qofolz"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-1baa01z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="m11.54 22.351.07.04.028.016a.76.76 0 0 0 .723 0l.028-.015.071-.041a16.975 16.975 0 0 0 1.144-.742 19.58 19.58 0 0 0 2.683-2.282c1.944-1.99 3.963-4.98 3.963-8.827a8.25 8.25 0 0 0-16.5 0c0 3.846 2.02 6.837 3.963 8.827a19.58 19.58 0 0 0 2.682 2.282 16.975 16.975 0 0 0 1.145.742ZM12 13.5a3 3 0 1 0 0-6 3 3 0 0 0 0 6Z"
                  fill-rule="evenodd"
                />
              </svg>
              <p
                class="chakra-text css-qrr9p8"
                title="Location"
              >
                Location
              </p>
            </div>
          </th>
          <th
            class="css-ykdcwi"
          >
            <div
              class="css-1qofolz"
            >
              <svg
                class="chakra-icon css-1baa01z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M372 160v-12A116.13 116.13 0 0 0 258.89 32h-5.78A116.13 116.13 0 0 0 140 148v12H52a4 4 0 0 0-4 4v300a16 16 0 0 0 16 16h384a16 16 0 0 0 16-16V164a4 4 0 0 0-4-4zm-40 0H180v-11c0-41.84 33.41-76.56 75.25-77A76.08 76.08 0 0 1 332 148z"
                />
              </svg>
              <p
                class="chakra-text css-qrr9p8"
                title="Company"
              >
                Company
              </p>
            </div>
          </th>
          <th
            class="css-ykdcwi"
          >
            <div
              class="css-1qofolz"
            >
              <svg
                class="chakra-icon css-1baa01z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M208,32H48A16,16,0,0,0,32,48V208a16,16,0,0,0,16,16H208a16,16,0,0,0,16-16V48A16,16,0,0,0,208,32Zm-34.34,77.66-56,56a8,8,0,0,1-11.32,0l-24-24a8,8,0,0,1,11.32-11.32L112,148.69l50.34-50.35a8,8,0,0,1,11.32,11.32Z"
                />
              </svg>
              <p
                class="chakra-text css-qrr9p8"
                title="Status"
              >
                Status
              </p>
            </div>
          </th>
          <th
            class="css-ykdcwi"
          >
            <div
              class="css-1qofolz"
            >
              <p
                class="chakra-text css-qrr9p8"
              />
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <td
            class="css-zgnwis"
          >
            <div
              class="css-k008qs"
            >
              <label
                class="chakra-checkbox css-1elxtiq"
                data-testid="row-checkbox"
              >
                <input
                  aria-label="Checkbox 1"
                  class="chakra-checkbox__input"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="checkbox"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="chakra-checkbox__control css-dnty2r"
                />
              </label>
            </div>
          </td>
          <td
            class="css-1cthrm"
          >
            <div
              class="css-felsgu"
            >
              <span
                class="chakra-avatar css-1ruhx4w"
              >
                <div
                  aria-label="John Doe"
                  class="chakra-avatar__initials css-0"
                  role="img"
                >
                  JD
                </div>
              </span>
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-m3mw4z"
                >
                  John Doe
                </p>
                <p
                  class="chakra-text css-1np5o4d"
                />
              </div>
            </div>
          </td>
          <td
            class="css-1cthrm"
          >
            123-456-7890
          </td>
          <td
            class="css-1cthrm"
          >
            New York, NY
          </td>
          <td
            class="css-1cthrm"
          >
            ABC Company
          </td>
          <td
            class="css-1cthrm"
          >
            <div
              class="css-1i6zq7o"
            >
              <svg
                fill="none"
                height="4"
                viewBox="0 0 4 4"
                width="4"
              >
                <circle
                  cx="2.4375"
                  cy="1.59375"
                  fill="#05C168"
                  r="1.5"
                />
              </svg>
              <p
                class="chakra-text css-1k5z805"
              >
                Online
              </p>
            </div>
          </td>
          <td
            class="css-1cthrm"
          >
            <div
              class="css-147dy5z"
            >
              <button
                aria-label="Edit"
                class="chakra-button css-50f1ws"
                data-testid="edit-button"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 20 20"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m2.695 14.762-1.262 3.155a.5.5 0 0 0 .65.65l3.155-1.262a4 4 0 0 0 1.343-.886L17.5 5.501a2.121 2.121 0 0 0-3-3L3.58 13.419a4 4 0 0 0-.885 1.343Z"
                  />
                </svg>
              </button>
              <button
                aria-label="Delete"
                class="chakra-button css-50f1ws"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 6V3C7 2.44772 7.44772 2 8 2H16C16.5523 2 17 2.44772 17 3V6H22V8H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V8H2V6H7ZM9 4V6H15V4H9Z"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
        <tr
          class="css-0"
        >
          <td
            class="css-zgnwis"
          >
            <div
              class="css-k008qs"
            >
              <label
                class="chakra-checkbox css-1elxtiq"
                data-testid="row-checkbox"
              >
                <input
                  aria-label="Checkbox 2"
                  class="chakra-checkbox__input"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="checkbox"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="chakra-checkbox__control css-dnty2r"
                />
              </label>
            </div>
          </td>
          <td
            class="css-1cthrm"
          >
            <div
              class="css-felsgu"
            >
              <span
                class="chakra-avatar css-1ruhx4w"
              >
                <div
                  aria-label="Jane Doe"
                  class="chakra-avatar__initials css-0"
                  role="img"
                >
                  JD
                </div>
              </span>
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-m3mw4z"
                >
                  Jane Doe
                </p>
                <p
                  class="chakra-text css-1np5o4d"
                />
              </div>
            </div>
          </td>
          <td
            class="css-1cthrm"
          >
            987-654-3210
          </td>
          <td
            class="css-1cthrm"
          >
            Los Angeles, CA
          </td>
          <td
            class="css-1cthrm"
          >
            XYZ Company
          </td>
          <td
            class="css-1cthrm"
          >
            <div
              class="css-x8nfwg"
            >
              <svg
                fill="none"
                height="4"
                viewBox="0 0 4 4"
                width="4"
              >
                <circle
                  cx="2.4375"
                  cy="1.59375"
                  fill="#AEB9E1"
                  r="1.5"
                />
              </svg>
              <p
                class="chakra-text css-d1xskd"
              >
                Offline
              </p>
            </div>
          </td>
          <td
            class="css-1cthrm"
          >
            <div
              class="css-147dy5z"
            >
              <button
                aria-label="Edit"
                class="chakra-button css-50f1ws"
                data-testid="edit-button"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 20 20"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m2.695 14.762-1.262 3.155a.5.5 0 0 0 .65.65l3.155-1.262a4 4 0 0 0 1.343-.886L17.5 5.501a2.121 2.121 0 0 0-3-3L3.58 13.419a4 4 0 0 0-.885 1.343Z"
                  />
                </svg>
              </button>
              <button
                aria-label="Delete"
                class="chakra-button css-50f1ws"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 6V3C7 2.44772 7.44772 2 8 2H16C16.5523 2 17 2.44772 17 3V6H22V8H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V8H2V6H7ZM9 4V6H15V4H9Z"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
